File: mobile/lib/features/import/parser.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';
import '../../data/models/highlight_type.dart';
import '../../data/models/parsed_highlight.dart';

/// Parser for Kindle "My Clippings.txt" format
class KindleClippingsParser {
  static const String _separator = '==========';

  /// Parse Kindle clippings file content
  List<ParsedHighlight> parse(String content) {
    final highlights = <ParsedHighlight>[];

    // Split by separator
    final entries = content.split(_separator);

    for (final entry in entries) {
      final trimmed = entry.trim();
      if (trimmed.isEmpty) continue;

      final highlight = _parseEntry(trimmed);
      if (highlight != null) {
        highlights.add(highlight);
      }
    }

    return highlights;
  }

  ParsedHighlight? _parseEntry(String entry) {
    final lines = entry.split('\n').map((l) => l.trim()).toList();

    // Need at least: title line, metadata line, empty line, content
    if (lines.length < 3) return null;

    // Line 1: Book title (Author)
    final (title, author) = _parseTitleLine(lines[0]);
    if (title.isEmpty) return null;

    // Line 2: Metadata (type, location, page, date)
    final (type, location, page, kindleDate) = _parseMetadataLine(lines[1]);

    // Content: everything after metadata, skipping empty lines at start
    var contentStart = 2;
    while (contentStart < lines.length && lines[contentStart].isEmpty) {
      contentStart++;
    }

    if (contentStart >= lines.length) return null;

    final content = lines.sublist(contentStart).join('\n').trim();
    if (content.isEmpty) return null;

    return ParsedHighlight(
      bookTitle: title,
      author: author,
      content: content,
      type: type,
      location: location,
      page: page,
      kindleDate: kindleDate,
    );
  }

  (String title, String? author) _parseTitleLine(String line) {
    // Format: "Book Title (Author Name)" or just "Book Title"
    final parenStart = line.lastIndexOf('(');
    final parenEnd = line.lastIndexOf(')');

    if (parenStart > 0 && parenEnd > parenStart) {
      final title = line.substring(0, parenStart).trim();
      final author = line.substring(parenStart + 1, parenEnd).trim();
      return (title, author.isEmpty ? null : author);
    }

    return (line.trim(), null);
  }

  (HighlightType type, String? location, int? page, DateTime? kindleDate)
      _parseMetadataLine(String line) {
    // Determine type
    final type =
        line.contains('Note') ? HighlightType.note : HighlightType.highlight;

    // Extract location: "Location 72-75" or "Location 72"
    final location = _extractPattern(line, 'Location ', [' |', '\n', '\r']);

    // Extract page: "page 5"
    final pageStr = _extractPattern(line, 'page ', [' |', '\n', '\r']);
    final page = pageStr != null ? int.tryParse(pageStr) : null;

    // Extract date: "Added on Monday, January 20, 2026 10:30:00 AM"
    final kindleDate = _parseKindleDate(line);

    return (type, location, page, kindleDate);
  }

  String? _extractPattern(String text, String start, List<String> ends) {
    final startIdx = text.indexOf(start);
    if (startIdx < 0) return null;

    final valueStart = startIdx + start.length;
    final remaining = text.substring(valueStart);

    var endIdx = remaining.length;
    for (final end in ends) {
      final idx = remaining.indexOf(end);
      if (idx >= 0 && idx < endIdx) {
        endIdx = idx;
      }
    }

    final value = remaining.substring(0, endIdx).trim();
    return value.isEmpty ? null : value;
  }

  DateTime? _parseKindleDate(String line) {
    // Pattern: "Added on <weekday>, <month> <day>, <year> <time>"
    final addedOnIdx = line.indexOf('Added on ');
    if (addedOnIdx < 0) return null;

    final dateStr = line.substring(addedOnIdx + 9).trim();

    // Try multiple date formats
    return _tryParseDate(dateStr);
  }

  DateTime? _tryParseDate(String dateStr) {
    // Remove weekday prefix if present (e.g., "Monday, ")
    final commaIdx = dateStr.indexOf(', ');
    final cleanDate =
        commaIdx > 0 ? dateStr.substring(commaIdx + 2).trim() : dateStr.trim();

    // Try common formats
    final patterns = [
      // January 20, 2026 10:30:00 AM (US format)
      RegExp(
          r'(\w+)\s+(\d{1,2}),\s+(\d{4})\s+(\d{1,2}):(\d{2}):(\d{2})\s*(AM|PM)?',
          caseSensitive: false),
      // 20 January 2026 10:30:00 (UK/European format)
      RegExp(
          r'(\d{1,2})\s+(\w+)\s+(\d{4})\s+(\d{1,2}):(\d{2}):(\d{2})\s*(AM|PM)?',
          caseSensitive: false),
    ];

    for (final pattern in patterns) {
      final match = pattern.firstMatch(cleanDate);
      if (match != null) {
        return _buildDateTime(match);
      }
    }

    return null;
  }

  DateTime? _buildDateTime(RegExpMatch match) {
    try {
      final groups = match.groups([1, 2, 3, 4, 5, 6, 7]);

      // Determine if first format (month name first) or second (day first)
      int day;
      int month;
      int year;

      final first = groups[0]!;
      if (int.tryParse(first) != null) {
        // Day first format: day, month name, year
        day = int.parse(first);
        month = _parseMonth(groups[1]!);
        year = int.parse(groups[2]!);
      } else {
        // Month first format: month name, day, year
        month = _parseMonth(first);
        day = int.parse(groups[1]!);
        year = int.parse(groups[2]!);
      }

      var hour = int.parse(groups[3]!);
      final minute = int.parse(groups[4]!);
      final second = int.parse(groups[5]!);

      // Handle AM/PM
      final ampm = groups[6];
      if (ampm != null) {
        if (ampm.toUpperCase() == 'PM' && hour < 12) {
          hour += 12;
        } else if (ampm.toUpperCase() == 'AM' && hour == 12) {
          hour = 0;
        }
      }

      return DateTime(year, month, day, hour, minute, second);
    } catch (_) {
      return null;
    }
  }

  int _parseMonth(String month) {
    const months = {
      'january': 1,
      'february': 2,
      'march': 3,
      'april': 4,
      'may': 5,
      'june': 6,
      'july': 7,
      'august': 8,
      'september': 9,
      'october': 10,
      'november': 11,
      'december': 12,
    };
    return months[month.toLowerCase()] ?? 1;
  }

  /// Generate content hash for duplicate detection
  static String generateContentHash(String bookTitle, String content) {
    final normalized = '${bookTitle.toLowerCase().trim()}|${content.toLowerCase().trim()}';
    final bytes = utf8.encode(normalized);
    final digest = sha256.convert(bytes);
    return digest.toString();
  }
}

File: desktop/src-tauri/src/kindle/parser.rs
// Kindle clippings file parser

use anyhow::Result;
use chrono::{DateTime, NaiveDateTime, Utc};
use once_cell::sync::Lazy;
use regex::Regex;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

/// Parsed highlight from Kindle clippings file
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParsedHighlight {
    pub book_title: String,
    pub author: Option<String>,
    pub content: String,
    pub highlight_type: HighlightType,
    pub location: Option<String>,
    pub page: Option<i32>,
    pub kindle_date: Option<DateTime<Utc>>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub enum HighlightType {
    Highlight,
    Note,
    Bookmark,
}

/// Parse Kindle clippings file content
pub fn parse_clippings(content: &str) -> Result<Vec<ParsedHighlight>> {
    let mut highlights = Vec::new();

    // Split by separator (==========)
    let entries: Vec<&str> = content.split("==========").collect();

    for entry in entries {
        let entry = entry.trim();
        if entry.is_empty() {
            continue;
        }

        if let Some(highlight) = parse_single_entry(entry) {
            highlights.push(highlight);
        }
    }

    Ok(highlights)
}

fn parse_single_entry(entry: &str) -> Option<ParsedHighlight> {
    let lines: Vec<&str> = entry.lines().collect();

    if lines.len() < 3 {
        return None;
    }

    // Line 1: Book title (Author)
    let (title, author) = parse_title_line(lines[0]);

    // Line 2: Metadata (type, location, page, date)
    let (highlight_type, location, page, kindle_date) = parse_metadata_line(lines[1]);

    // Line 3+: Content (skip empty line after metadata)
    let content_start = if lines.get(2).map(|s| s.is_empty()).unwrap_or(false) { 3 } else { 2 };
    let content: String = lines[content_start..].join("\n").trim().to_string();

    if content.is_empty() {
        return None;
    }

    Some(ParsedHighlight {
        book_title: title,
        author,
        content,
        highlight_type,
        location,
        page,
        kindle_date,
    })
}

fn parse_title_line(line: &str) -> (String, Option<String>) {
    // Format: "Book Title (Author Name)"
    if let Some(paren_start) = line.rfind('(') {
        if let Some(paren_end) = line.rfind(')') {
            let title = line[..paren_start].trim().to_string();
            let author = line[paren_start + 1..paren_end].trim().to_string();
            return (title, Some(author));
        }
    }
    (line.trim().to_string(), None)
}

fn parse_metadata_line(line: &str) -> (HighlightType, Option<String>, Option<i32>, Option<DateTime<Utc>>) {
    let highlight_type = if line.contains("Note") {
        HighlightType::Note
    } else {
        HighlightType::Highlight
    };

    // Extract location
    let location = extract_pattern(line, "Location ", " |")
        .or_else(|| extract_pattern(line, "Location ", "\n"));

    // Extract page
    let page = extract_pattern(line, "page ", " |")
        .or_else(|| extract_pattern(line, "page ", "\n"))
        .and_then(|s| s.parse().ok());

    // TODO: Parse date (format varies by Kindle language settings)
    let kindle_date = None;

    (highlight_type, location, page, kindle_date)
}

fn extract_pattern(text: &str, start: &str, end: &str) -> Option<String> {
    let start_idx = text.find(start)?;
    let value_start = start_idx + start.len();
    let remaining = &text[value_start..];
    let end_idx = remaining.find(end).unwrap_or(remaining.len());
    Some(remaining[..end_idx].trim().to_string())
}

// ============================================
// Regex patterns for robust parsing
// ============================================

/// Regex patterns for parsing Kindle clippings
pub mod patterns {
    use super::*;

    /// Pattern for title with author: "Book Title (Author Name)"
    pub static TITLE_AUTHOR: Lazy<Regex> = Lazy::new(|| {
        Regex::new(r"^(.+?)\s*\(([^)]+)\)\s*$").expect("Invalid title regex")
    });

    /// Pattern for highlight/note type
    pub static ENTRY_TYPE: Lazy<Regex> = Lazy::new(|| {
        Regex::new(r"Your (Highlight|Note|Bookmark)").expect("Invalid type regex")
    });

    /// Pattern for page number
    pub static PAGE_NUMBER: Lazy<Regex> = Lazy::new(|| {
        Regex::new(r"page\s+(\d+)").expect("Invalid page regex")
    });

    /// Pattern for location
    pub static LOCATION: Lazy<Regex> = Lazy::new(|| {
        Regex::new(r"Location\s+([\d\-]+)").expect("Invalid location regex")
    });

    /// US date format: "Monday, January 15, 2024 12:30:45 PM"
    pub static DATE_US: Lazy<Regex> = Lazy::new(|| {
        Regex::new(
            r"Added on \w+,\s+(\w+)\s+(\d{1,2}),\s+(\d{4})\s+(\d{1,2}):(\d{2}):(\d{2})\s*(AM|PM)"
        ).expect("Invalid US date regex")
    });

    /// UK date format: "Monday, 15 January 2024 12:30:45"
    pub static DATE_UK: Lazy<Regex> = Lazy::new(|| {
        Regex::new(
            r"Added on \w+,\s+(\d{1,2})\s+(\w+)\s+(\d{4})\s+(\d{1,2}):(\d{2}):(\d{2})"
        ).expect("Invalid UK date regex")
    });

    /// Parse title and author using regex
    pub fn parse_title_regex(line: &str) -> (String, Option<String>) {
        if let Some(captures) = TITLE_AUTHOR.captures(line) {
            let title = captures.get(1).map(|m| m.as_str().trim().to_string()).unwrap_or_default();
            let author = captures.get(2).map(|m| m.as_str().trim().to_string());
            return (title, author);
        }
        (line.trim().to_string(), None)
    }

    /// Parse highlight type from metadata line
    pub fn parse_type_regex(line: &str) -> HighlightType {
        if let Some(captures) = ENTRY_TYPE.captures(line) {
            match captures.get(1).map(|m| m.as_str()) {
                Some("Note") => return HighlightType::Note,
                Some("Bookmark") => return HighlightType::Bookmark,
                _ => return HighlightType::Highlight,
            }
        }
        HighlightType::Highlight
    }

    /// Parse page number from metadata line
    pub fn parse_page_regex(line: &str) -> Option<i32> {
        PAGE_NUMBER.captures(line)
            .and_then(|c| c.get(1))
            .and_then(|m| m.as_str().parse().ok())
    }

    /// Parse location from metadata line
    pub fn parse_location_regex(line: &str) -> Option<String> {
        LOCATION.captures(line)
            .and_then(|c| c.get(1))
            .map(|m| m.as_str().to_string())
    }

    /// Parse date from metadata line (handles US and UK formats)
    pub fn parse_date_regex(line: &str) -> Option<DateTime<Utc>> {
        // Try US format first
        if let Some(captures) = DATE_US.captures(line) {
            return parse_us_date(&captures);
        }
        // Try UK format
        if let Some(captures) = DATE_UK.captures(line) {
            return parse_uk_date(&captures);
        }
        None
    }

    fn parse_us_date(captures: &regex::Captures) -> Option<DateTime<Utc>> {
        let month_str = captures.get(1)?.as_str();
        let day: u32 = captures.get(2)?.as_str().parse().ok()?;
        let year: i32 = captures.get(3)?.as_str().parse().ok()?;
        let mut hour: u32 = captures.get(4)?.as_str().parse().ok()?;
        let minute: u32 = captures.get(5)?.as_str().parse().ok()?;
        let second: u32 = captures.get(6)?.as_str().parse().ok()?;
        let am_pm = captures.get(7)?.as_str();

        // Convert to 24-hour format
        if am_pm == "PM" && hour != 12 {
            hour += 12;
        } else if am_pm == "AM" && hour == 12 {
            hour = 0;
        }

        let month = month_to_number(month_str)?;
        let naive = NaiveDateTime::parse_from_str(
            &format!("{:04}-{:02}-{:02} {:02}:{:02}:{:02}", year, month, day, hour, minute, second),
            "%Y-%m-%d %H:%M:%S"
        ).ok()?;

        Some(DateTime::from_naive_utc_and_offset(naive, Utc))
    }

    fn parse_uk_date(captures: &regex::Captures) -> Option<DateTime<Utc>> {
        let day: u32 = captures.get(1)?.as_str().parse().ok()?;
        let month_str = captures.get(2)?.as_str();
        let year: i32 = captures.get(3)?.as_str().parse().ok()?;
        let hour: u32 = captures.get(4)?.as_str().parse().ok()?;
        let minute: u32 = captures.get(5)?.as_str().parse().ok()?;
        let second: u32 = captures.get(6)?.as_str().parse().ok()?;

        let month = month_to_number(month_str)?;
        let naive = NaiveDateTime::parse_from_str(
            &format!("{:04}-{:02}-{:02} {:02}:{:02}:{:02}", year, month, day, hour, minute, second),
            "%Y-%m-%d %H:%M:%S"
        ).ok()?;

        Some(DateTime::from_naive_utc_and_offset(naive, Utc))
    }

    fn month_to_number(month: &str) -> Option<u32> {
        match month.to_lowercase().as_str() {
            "january" | "jan" => Some(1),
            "february" | "feb" => Some(2),
            "march" | "mar" => Some(3),
            "april" | "apr" => Some(4),
            "may" => Some(5),
            "june" | "jun" => Some(6),
            "july" | "jul" => Some(7),
            "august" | "aug" => Some(8),
            "september" | "sep" => Some(9),
            "october" | "oct" => Some(10),
            "november" | "nov" => Some(11),
            "december" | "dec" => Some(12),
            _ => None,
        }
    }
}

/// Generate SHA-256 content hash for duplicate detection
pub fn generate_content_hash(book_title: &str, content: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(book_title.as_bytes());
    hasher.update(content.as_bytes());
    hex::encode(hasher.finalize())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_title_with_author() {
        let (title, author) = parse_title_line("The Great Gatsby (F. Scott Fitzgerald)");
        assert_eq!(title, "The Great Gatsby");
        assert_eq!(author, Some("F. Scott Fitzgerald".to_string()));
    }

    #[test]
    fn test_parse_title_without_author() {
        let (title, author) = parse_title_line("Unknown Book");
        assert_eq!(title, "Unknown Book");
        assert_eq!(author, None);
    }

    #[test]
    fn test_parse_clippings() {
        let content = r#"The Great Gatsby (F. Scott Fitzgerald)
- Your Highlight on page 5 | Location 72-75 | Added on Monday, January 20, 2026 10:30:00 AM

In my younger and more vulnerable years my father gave me some advice.
==========
1984 (George Orwell)
- Your Note on page 50 | Location 750 | Added on Tuesday, January 21, 2026 9:30:00 AM

This reminds me of modern surveillance.
=========="#;

        let highlights = parse_clippings(content).unwrap();
        assert_eq!(highlights.len(), 2);
        assert_eq!(highlights[0].book_title, "The Great Gatsby");
        assert_eq!(highlights[0].highlight_type, HighlightType::Highlight);
        assert_eq!(highlights[1].book_title, "1984");
        assert_eq!(highlights[1].highlight_type, HighlightType::Note);
    }
}
